python/lib/modeldata/dmod/modeldata/data/bmi_init_config_auto_generator.py

```diff
@@ -0,0 +1,475 @@
+        # > SOILCOLOR is hard-coded as 4 in module_sf_noahmpdrv.F in the current release of HRLDAS. SOILCOLOR is used to select the albedo values for dry and saturated soil.
+        # NOTE: it appears that the soil color indexes into the ALBSAT_VIS, ALBSAT_NIR, ALBDRY_VIS, ALBDRY_NIR tables
+        # https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/parameters/MPTABLE.TBL#L328-L331
+        # here is the indexing
+        # https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/src/ParametersType.f90#L303-L306
+        # NOTE: looks like for HRLDAS this is 4
+        soilcolor: int = 4
+        structure = Structure(
+            isltyp=isltyp,
+            nsoil=nsoil,
+            nsnow=nsnow,
+            nveg=nveg,
+            vegtyp=vegtyp,
+            # crop type (SET TO 0, no crops currently supported)
+            # source: https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/src/NamelistRead.f90#L36
+            croptype=croptype,
+            sfctyp=sfctyp,
+            soilcolor=soilcolor,
+        )
+        self.data["structure"] = structure
+
+    def visit(self, hook_provider: "HookProvider") -> None:
+        hook_provider.provide_hydrofabric_linked_data(self)
+
+        self._v2_defaults()
+
+    def build(self) -> BaseModel:
+        return NoahOWPConfig(**self.data)
+
+
+class BmiGenDivideIdHookObject:
+    def __init__(self):
+        self.__divide_id: Union[str, None] = None
+
+    def hydrofabric_hook(self, version: str, divide_id: str, data: Dict[str, Any]) -> None:
+        self.__divide_id = divide_id
+
+    def visit(self, hook_provider) -> None:
+        hook_provider.provide_hydrofabric_data(self)
+
+    def divide_id(self) -> Union[str, None]:
+        return self.__divide_id
+
+
+# TODO: figure out how to handle noah owp
+class BmiInitConfigAutoGenerator:
+    """
+    A tool for automatically generating required BMI initialization configs for a realization config and hydrofabric.
+
+    A helper tool for generating BMI init configs as needed for a given realization config.  It also requires data from
+    a hydrofabric dataset, not only because a realization config only has real meaning in the context of a particular
+    hydrofabric (i.e., otherwise, what exactly is, for example, ``cat-115`` referring to?), but because the hydrofabric
+    dataset contains model attributes data from which an applicable BMI init config can be generated.
+
+    This type strategically applies classes in the ``ngen.config_gen`` package to perform the actual generation of
+    configs.
+
+    See https://www.lynker-spatial.com/copyright.html for license details on hydrofabric data.
+    """
+
+    _module_to_model_map: Dict[str, Any] = {"CFE": Cfe, "PET": Pet}
+    """ Map of config strings to builders, for modules with builders than can be easily init with no more info. """
+    _no_init_config_modules = {"SLOTH"}
```

Comment:
Similarly to the above, I didn't see a scenario when we could avoid excluding `SLOTH`.  So hard-coding the exclusion seemed appropriate.

I suppose (and again, this lines up with the earlier param) this all assumes that an unrecognized `model_type_name` in the realization config is invalid, rather than something the user wants ignored.  I feel like that is more appropriate in this context:  fail immediately, instead of silently trusting that the user didn't make a mistake (e.g., a typo) and turning any actual errors here into failures at a later step.  But I'm open to discussing that further.

Reply:
Same thinking as above. It will be easier to update service code rather than library code.